runtime/src/accounts_db.rs

```diff
@@ -2715,12 +2697,53 @@ impl AccountsDB {
         }
     }
 
-    fn store_with_hashes(&self, slot: Slot, accounts: &[(&Pubkey, &Account)], hashes: &[Hash]) {
+    fn store_accounts_default<'a>(
+        &'a self,
+        slot: Slot,
+        accounts: &[(&Pubkey, &Account)],
+        hashes: &[Hash],
+    ) {
+        self.store_accounts_custom(
+            slot,
+            accounts,
+            hashes,
+            None::<StorageFinder>,
+            None::<Box<dyn Iterator<Item = u64>>>,
+        );
+    }
+
+    fn store_accounts_custom<'a>(
+        &'a self,
+        slot: Slot,
+        accounts: &[(&Pubkey, &Account)],
+        hashes: &[Hash],
+        storage_finder: Option<StorageFinder<'a>>,
+        write_version_producer: Option<Box<dyn Iterator<Item = u64>>>,
+    ) -> StoreAccountsTiming {
+        let storage_finder: StorageFinder<'a> = storage_finder
+            .unwrap_or_else(|| Box::new(move |slot, size| self.find_storage_candidate(slot, size)));
+
+        let write_version_producer: Box<dyn Iterator<Item = u64>> = write_version_producer
+            .unwrap_or_else(|| {
+                let mut current_version = self.bulk_assign_write_version(accounts.len());
+                Box::new(std::iter::from_fn(move || {
+                    let ret = current_version;
```

Comment:
@carllin Do you like fun rust ride?  :) Could you try `impl Fn(...)` instead of `Box<dyn Fn(...)>` like I played at #13345. What I'm worrying is that this closure will be called inside a tight loop. With `Box`, I think this incurrs dynamic dispatch for each invocation. I dunno how that affects the performance. Or rustc is smart enough to optimize away it. But I'd like to avoid that altogether if could by using `impl Fn` (which does static diapatch; also no heap allocation by not using `Box`). Abstraction with zero runtime cost. Beware, compiler- lang- support was rather edgy for `impl Fn`...

Reply:
hehe @ryoqun I actually spent quite some time tying to avoid this `Box` as well, and I think it's tradeoff for code size (alternative is creating a case for every one of the combinations). 

The problem with using `impl Fn(...)` is that the `write_version_producer` needs to be able to infer a type at compile time, so something like `write_version_producer` cannot hold two different types here (different in the `if` and `else` branch).

Reply:
I see. Thanks for trying so hard and sharing the result. :)